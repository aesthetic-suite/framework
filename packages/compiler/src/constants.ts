import { ScaleType, BreakpointListConfig } from './types';

export const NAME_PATTERN = /^[-a-z0-9]+$/giu;

export const CONFIG_FOLDER = '.aesthetic';

export const BRAND_FILE = 'brand.yaml';

export const LANGUAGE_FILE = 'language.yaml';

export const THEMES_FILE = 'themes.yaml';

export const DEFAULT_BREAKPOINTS: BreakpointListConfig = [640, 960, 1280, 1600, 1920];

export const DEFAULT_UNIT = 8;

export const FONT_FAMILIES = {
  'android-monospace': 'monospace',
  'android-system': 'sans-serif',
  'ios-monospace': 'Menlo-Regular',
  'ios-system': 'SF Pro',
  'web-monospace':
    'Menlo, Monaco, Consolas, "Lucida Console", "Liberation Mono", "Courier New", monospace',
  'web-system':
    'system-ui, -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, "Noto Sans", sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji"',
};

// https://learnui.design/blog/ultimate-guide-font-sizes-ui-design.html
export const PLATFORM_CONFIGS = {
  android: {
    baseFontSize: 14,
    fontUnit: 'sp',
    spacingUnit: 'dp',
  },
  ios: {
    baseFontSize: 17,
    fontUnit: 'pt',
    spacingUnit: 'pt', // ???
  },
  web: {
    baseFontSize: 16,
    fontUnit: 'px',
    spacingUnit: 'rem',
  },
};

// https://type-scale.com
export const SCALES: { [K in ScaleType]: number } = {
  'augmented-fourth': 1.414,
  'double-octave': 4,
  'golden-ratio': 1.618,
  'golden-section': 1.618,
  'major-second': 1.125,
  'major-third': 1.25,
  'major-sixth': 1.667,
  'major-seventh': 1.875,
  'major-tenth': 2.5,
  'major-eleventh': 2.667,
  'major-twelfth': 3,
  'minor-second': 1.067,
  'minor-third': 1.2,
  'minor-sixth': 1.6,
  'minor-seventh': 1.778,
  octave: 2,
  'perfect-fifth': 1.5,
  'perfect-fourth': 1.333,
};
