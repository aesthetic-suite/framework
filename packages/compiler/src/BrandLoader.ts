import optimal, { string } from 'optimal';
import { DeepPartial } from '@aesthetic/system';
import { Path } from '@boost/common';
import { BRAND_FILE, NAME_PATTERN } from './constants';
import Loader from './Loader';
import { BrandConfigFile } from './types';

export default class BrandLoader extends Loader<BrandConfigFile> {
	getFileName() {
		return BRAND_FILE;
	}

	validate(config: DeepPartial<BrandConfigFile>, filePath: Path): BrandConfigFile {
		return optimal(
			config,
			{
				extends: string(),
				name: string().notEmpty().match(NAME_PATTERN),
			},
			{ file: filePath.path() },
		);
	}
}
