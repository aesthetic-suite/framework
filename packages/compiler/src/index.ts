/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import Compiler from './Compiler';
import BrandLoader from './BrandLoader';
import LanguageLoader from './LanguageLoader';
import ThemesLoader from './ThemesLoader';

export * from './constants';
export * from './types';

export { Compiler, BrandLoader, LanguageLoader, ThemesLoader };
