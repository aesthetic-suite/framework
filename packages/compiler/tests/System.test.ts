import { Path } from '@boost/common';
import { FONT_FAMILIES, SystemOptions } from '../src';
import { LanguageLoader } from '../src/LanguageLoader';
import { SystemDesign } from '../src/SystemDesign';
import { ThemesLoader } from '../src/ThemesLoader';

describe('System', () => {
	const options: SystemOptions = {
		format: 'web-tsx',
		platform: 'web',
	};

	describe('breakpoints', () => {
		it('handles mobile first', () => {
			const config = new LanguageLoader('web').load(
				new Path(__dirname, './__fixtures__/system-fixed/language.yaml'),
			);

			const design = new SystemDesign('test', config, options);

			expect(design.template.breakpoint).toEqual({
				xs: {
					queryConditions: [['min-width', 640]],
					querySize: 640,
					rootLineHeight: 1.333_75,
					rootTextSize: 14.937_999_999_999_999,
				},
				sm: {
					queryConditions: [['min-width', 960]],
					querySize: 960,
					rootLineHeight: 1.423_111_249_999_999_8,
					rootTextSize: 15.938_845_999_999_998,
				},
				md: {
					queryConditions: [['min-width', 1280]],
					querySize: 1280,
					rootLineHeight: 1.518_459_703_749_999_8,
					rootTextSize: 17.006_748_681_999_998,
				},
				lg: {
					queryConditions: [['min-width', 1600]],
					querySize: 1600,
					rootLineHeight: 1.620_196_503_901_249_8,
					rootTextSize: 18.146_200_843_693_997,
				},
				xl: {
					queryConditions: [['min-width', 1920]],
					querySize: 1920,
					rootLineHeight: 1.728_749_669_662_633_4,
					rootTextSize: 19.361_996_300_221_495,
				},
			});
		});
	});

	describe('typography', () => {
		it('sets each font', () => {
			const config = new LanguageLoader('web').load(
				new Path(__dirname, './__fixtures__/fonts.yaml'),
			);

			const design = new SystemDesign('test', config, options);

			expect(design.template.typography.font).toEqual({
				text: 'Roboto, sans-serif',
				heading: `Droid, ${FONT_FAMILIES['web-system']}`,
				monospace: '"Lucida Console", Monaco, monospace',
				system: FONT_FAMILIES['web-system'],
				locale: {
					ja_JP: 'YuGothic, "Meiryo UI", Meiryo, Osaka, Tahoma, Arial, sans-serif',
					th_TH: '"Leelawadee UI Regular", "Kmer UI", Tahoma, Arial, sans-serif',
				},
			});
		});

		it('sets heading and text font based on "system"', () => {
			const config = new LanguageLoader('web').load(
				new Path(__dirname, './__fixtures__/system-font.yaml'),
			);

			const design = new SystemDesign('test', config, options);

			expect(design.template.typography.font.heading).toBe(FONT_FAMILIES['web-system']);
			expect(design.template.typography.font.text).toBe(FONT_FAMILIES['web-system']);
		});

		it('sets explicit heading and text font', () => {
			const config = new LanguageLoader('web').load(
				new Path(__dirname, './__fixtures__/explicit-font.yaml'),
			);

			const design = new SystemDesign('test', config, options);

			expect(design.template.typography.font.heading).toBe('Roboto, "Open Sans"');
			expect(design.template.typography.font.text).toBe('Roboto, "Open Sans"');
		});
	});

	describe('themes', () => {
		it('loads a theme and all palette colors', () => {
			const config = new LanguageLoader('web').load(
				new Path(__dirname, './__fixtures__/colors.yaml'),
			);

			const themes = new ThemesLoader(config.colors).load(
				new Path(__dirname, './__fixtures__/themes/themes.yaml'),
			);

			const design = new SystemDesign('test', config, options);
			const theme = design.createTheme('default', themes.themes.default);

			expect(theme.template).toMatchSnapshot();
		});

		it('supports theme extending', () => {
			const config = new LanguageLoader('web').load(
				new Path(__dirname, './__fixtures__/colors.yaml'),
			);

			const themes = new ThemesLoader(config.colors).load(
				new Path(__dirname, './__fixtures__/themes/themes.yaml'),
			);

			const design = new SystemDesign('test', config, options);
			const base = design.createTheme('default', themes.themes.default);
			const other = base.extend('other', themes.themes.other, 'default');

			expect(other.template).toMatchSnapshot();
		});

		it('supports palettes with different combinations of color and shade settings', () => {
			const config = new LanguageLoader('web').load(
				new Path(__dirname, './__fixtures__/colors.yaml'),
			);

			const themes = new ThemesLoader(config.colors).load(
				new Path(__dirname, './__fixtures__/themes/theme-palette-settings.yaml'),
			);

			const design = new SystemDesign('test', config, options);
			const theme = design.createTheme('default', themes.themes.default);

			expect(theme.template).toMatchSnapshot();
		});

		it('errors when theme doesnt implement the defined colors', () => {
			expect(() =>
				new ThemesLoader(['black', 'white']).load(
					new Path(__dirname, './__fixtures__/themes/missing-theme-color.yaml'),
				),
			).toThrow(
				'Invalid field "themes.default.colors". Theme has not implemented the following colors: black, white',
			);
		});

		it('errors when theme implements an unknown colors', () => {
			expect(() =>
				new ThemesLoader(['black']).load(
					new Path(__dirname, './__fixtures__/themes/unknown-theme-color.yaml'),
				),
			).toThrow('Invalid field "themes.default.colors". Theme is using unknown colors: white');
		});

		it('errors when theme implements too many color ranges', () => {
			expect(() =>
				new ThemesLoader(['black']).load(
					new Path(__dirname, './__fixtures__/themes/invalid-palette-range.yaml'),
				),
			).toThrow('Unknown "themes.default.colors.black" fields: 100.');
		});

		it('errors when theme palette references an invalid color', () => {
			expect(() =>
				new ThemesLoader(['black']).load(
					new Path(__dirname, './__fixtures__/themes/invalid-palette-color.yaml'),
				),
			).toThrowErrorMatchingSnapshot();
		});

		it('errors when theme palette references an invalid color shade', () => {
			expect(() =>
				new ThemesLoader(['black']).load(
					new Path(__dirname, './__fixtures__/themes/invalid-palette-color-reference.yaml'),
				),
			).toThrowErrorMatchingSnapshot();
		});
	});
});
