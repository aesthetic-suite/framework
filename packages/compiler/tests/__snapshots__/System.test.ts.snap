// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`System themes errors when theme palette references an invalid color 1`] = `
"Invalid field \\"themes.default.palettes.brand\\". Received object but must be a union of: string, shape
 - Invalid field \\"themes.default.palettes.brand.color\\". Invalid color \\"red\\"."
`;

exports[`System themes errors when theme palette references an invalid color shade 1`] = `
"Invalid field \\"themes.default.palettes.brand\\". Received object but must be a union of: string, shape
 - Invalid field \\"themes.default.palettes.brand.bg\\". Received object but must be a union of: string, shape
 - Invalid field \\"themes.default.palettes.brand.bg.base\\". String does not match. (pattern \\"^[a-z0-9]+(.(00|10|20|30|40|50|60|70|80|90))?$\\")"
`;

exports[`System themes loads a theme and all palette colors 1`] = `
Object {
  "palette": Object {
    "brand": Object {
      "bg": Object {
        "base": "#004",
        "disabled": "#003",
        "focused": "#005",
        "hovered": "#006",
        "selected": "#005",
      },
      "color": Object {
        "00": "#000",
        "10": "#001",
        "20": "#002",
        "30": "#003",
        "40": "#004",
        "50": "#005",
        "60": "#006",
        "70": "#007",
        "80": "#008",
        "90": "#009",
      },
      "fg": Object {
        "base": "#004",
        "disabled": "#003",
        "focused": "#005",
        "hovered": "#006",
        "selected": "#005",
      },
      "text": "#008",
    },
    "danger": Object {
      "bg": Object {
        "base": "#BE4",
        "disabled": "#BE3",
        "focused": "#BE5",
        "hovered": "#BE6",
        "selected": "#BE5",
      },
      "color": Object {
        "00": "#BE0",
        "10": "#BE1",
        "20": "#BE2",
        "30": "#BE3",
        "40": "#BE4",
        "50": "#BE5",
        "60": "#BE6",
        "70": "#BE7",
        "80": "#BE8",
        "90": "#BE9",
      },
      "fg": Object {
        "base": "#FF4",
        "disabled": "#FF3",
        "focused": "#FF5",
        "hovered": "#FF6",
        "selected": "#FF5",
      },
      "text": "#BE4",
    },
    "failure": Object {
      "bg": Object {
        "base": "#BE4",
        "disabled": "#BE3",
        "focused": "#BE5",
        "hovered": "#BE6",
        "selected": "#BE5",
      },
      "color": Object {
        "00": "#BE0",
        "10": "#BE1",
        "20": "#BE2",
        "30": "#BE3",
        "40": "#BE4",
        "50": "#BE5",
        "60": "#BE6",
        "70": "#BE7",
        "80": "#BE8",
        "90": "#BE9",
      },
      "fg": Object {
        "base": "#FF4",
        "disabled": "#FF3",
        "focused": "#FF5",
        "hovered": "#FF6",
        "selected": "#FF5",
      },
      "text": "#BE0",
    },
    "info": Object {
      "bg": Object {
        "base": "#FF4",
        "disabled": "#FF3",
        "focused": "#FF5",
        "hovered": "#FF6",
        "selected": "#FF5",
      },
      "color": Object {
        "00": "#FF0",
        "10": "#FF1",
        "20": "#FF2",
        "30": "#FF3",
        "40": "#FF4",
        "50": "#FF5",
        "60": "#FF6",
        "70": "#FF7",
        "80": "#FF8",
        "90": "#FF9",
      },
      "fg": Object {
        "base": "#004",
        "disabled": "#003",
        "focused": "#005",
        "hovered": "#006",
        "selected": "#005",
      },
      "text": "#FF8",
    },
    "muted": Object {
      "bg": Object {
        "base": "#004",
        "disabled": "#003",
        "focused": "#005",
        "hovered": "#006",
        "selected": "#005",
      },
      "color": Object {
        "00": "#000",
        "10": "#001",
        "20": "#002",
        "30": "#003",
        "40": "#004",
        "50": "#005",
        "60": "#006",
        "70": "#007",
        "80": "#008",
        "90": "#009",
      },
      "fg": Object {
        "base": "#FF4",
        "disabled": "#FF3",
        "focused": "#FF5",
        "hovered": "#FF6",
        "selected": "#FF5",
      },
      "text": "#008",
    },
    "neutral": Object {
      "bg": Object {
        "base": "#EE4",
        "disabled": "#EE3",
        "focused": "#EE5",
        "hovered": "#EE6",
        "selected": "#EE5",
      },
      "color": Object {
        "00": "#EE0",
        "10": "#EE1",
        "20": "#EE2",
        "30": "#EE3",
        "40": "#EE4",
        "50": "#EE5",
        "60": "#EE6",
        "70": "#EE7",
        "80": "#EE8",
        "90": "#EE9",
      },
      "fg": Object {
        "base": "#FF0",
        "disabled": "#FF1",
        "focused": "#FF2",
        "hovered": "#FF3",
        "selected": "#FF4",
      },
      "text": "#EE8",
    },
    "positive": Object {
      "bg": Object {
        "base": "#EE4",
        "disabled": "#EE3",
        "focused": "#EE5",
        "hovered": "#EE6",
        "selected": "#EE5",
      },
      "color": Object {
        "00": "#EE0",
        "10": "#EE1",
        "20": "#EE2",
        "30": "#EE3",
        "40": "#EE4",
        "50": "#EE5",
        "60": "#EE6",
        "70": "#EE7",
        "80": "#EE8",
        "90": "#EE9",
      },
      "fg": Object {
        "base": "#FF4",
        "disabled": "#FF3",
        "focused": "#FF5",
        "hovered": "#FF6",
        "selected": "#FF5",
      },
      "text": "#EE9",
    },
    "primary": Object {
      "bg": Object {
        "base": "#BE0",
        "disabled": "#BE1",
        "focused": "#005",
        "hovered": "#BE2",
        "selected": "#005",
      },
      "color": Object {
        "00": "#BE0",
        "10": "#BE1",
        "20": "#BE2",
        "30": "#BE3",
        "40": "#BE4",
        "50": "#BE5",
        "60": "#BE6",
        "70": "#BE7",
        "80": "#BE8",
        "90": "#BE9",
      },
      "fg": Object {
        "base": "#FF9",
        "disabled": "#003",
        "focused": "#FF8",
        "hovered": "#006",
        "selected": "#FF7",
      },
      "text": "#BE8",
    },
    "secondary": Object {
      "bg": Object {
        "base": "#ED4",
        "disabled": "#ED3",
        "focused": "#ED5",
        "hovered": "#ED6",
        "selected": "#ED5",
      },
      "color": Object {
        "00": "#ED0",
        "10": "#ED1",
        "20": "#ED2",
        "30": "#ED3",
        "40": "#ED4",
        "50": "#ED5",
        "60": "#ED6",
        "70": "#ED7",
        "80": "#ED8",
        "90": "#ED9",
      },
      "fg": Object {
        "base": "#FF4",
        "disabled": "#FF3",
        "focused": "#FF5",
        "hovered": "#FF6",
        "selected": "#FF5",
      },
      "text": "#ED8",
    },
    "tertiary": Object {
      "bg": Object {
        "base": "#BE4",
        "disabled": "#BE3",
        "focused": "#BE5",
        "hovered": "#BE6",
        "selected": "#BE5",
      },
      "color": Object {
        "00": "#BE0",
        "10": "#BE1",
        "20": "#BE2",
        "30": "#BE3",
        "40": "#BE4",
        "50": "#BE5",
        "60": "#BE6",
        "70": "#BE7",
        "80": "#BE8",
        "90": "#BE9",
      },
      "fg": Object {
        "base": "#BE4",
        "disabled": "#BE3",
        "focused": "#BE5",
        "hovered": "#BE6",
        "selected": "#BE5",
      },
      "text": "#BE8",
    },
    "warning": Object {
      "bg": Object {
        "base": "#ED4",
        "disabled": "#ED3",
        "focused": "#ED5",
        "hovered": "#ED6",
        "selected": "#ED5",
      },
      "color": Object {
        "00": "#ED0",
        "10": "#ED1",
        "20": "#ED2",
        "30": "#ED3",
        "40": "#ED4",
        "50": "#ED5",
        "60": "#ED6",
        "70": "#ED7",
        "80": "#ED8",
        "90": "#ED9",
      },
      "fg": Object {
        "base": "#FF9",
        "disabled": "#FF9",
        "focused": "#FF9",
        "hovered": "#FF9",
        "selected": "#FF9",
      },
      "text": "#ED8",
    },
  },
}
`;

exports[`System themes supports palettes with different combinations of color and shade settings 1`] = `
Object {
  "palette": Object {
    "brand": Object {
      "bg": Object {
        "base": "#004",
        "disabled": "#003",
        "focused": "#005",
        "hovered": "#006",
        "selected": "#005",
      },
      "color": Object {
        "00": "#000",
        "10": "#001",
        "20": "#002",
        "30": "#003",
        "40": "#004",
        "50": "#005",
        "60": "#006",
        "70": "#007",
        "80": "#008",
        "90": "#009",
      },
      "fg": Object {
        "base": "#004",
        "disabled": "#003",
        "focused": "#005",
        "hovered": "#006",
        "selected": "#005",
      },
      "text": "#008",
    },
    "danger": Object {
      "bg": Object {
        "base": "#BE3",
        "disabled": "#003",
        "focused": "#005",
        "hovered": "#006",
        "selected": "#005",
      },
      "color": Object {
        "00": "#000",
        "10": "#001",
        "20": "#002",
        "30": "#003",
        "40": "#004",
        "50": "#005",
        "60": "#006",
        "70": "#007",
        "80": "#008",
        "90": "#009",
      },
      "fg": Object {
        "base": "#EE4",
        "disabled": "#003",
        "focused": "#005",
        "hovered": "#006",
        "selected": "#005",
      },
      "text": "#ED5",
    },
    "failure": Object {
      "bg": Object {
        "base": "#EE4",
        "disabled": "#FF0",
        "focused": "#ED3",
        "hovered": "#BE2",
        "selected": "#001",
      },
      "color": Object {
        "00": "#000",
        "10": "#001",
        "20": "#002",
        "30": "#003",
        "40": "#004",
        "50": "#005",
        "60": "#006",
        "70": "#007",
        "80": "#008",
        "90": "#009",
      },
      "fg": Object {
        "base": "#ED9",
        "disabled": "#005",
        "focused": "#BE8",
        "hovered": "#EE7",
        "selected": "#FF6",
      },
      "text": "#FF0",
    },
    "info": Object {
      "bg": Object {
        "base": "#FF9",
        "disabled": "#FF0",
        "focused": "#FF8",
        "hovered": "#FF7",
        "selected": "#FF6",
      },
      "color": Object {
        "00": "#FF0",
        "10": "#FF1",
        "20": "#FF2",
        "30": "#FF3",
        "40": "#FF4",
        "50": "#FF5",
        "60": "#FF6",
        "70": "#FF7",
        "80": "#FF8",
        "90": "#FF9",
      },
      "fg": Object {
        "base": "#FF4",
        "disabled": "#FF3",
        "focused": "#FF5",
        "hovered": "#FF6",
        "selected": "#FF5",
      },
      "text": "#FF8",
    },
    "muted": Object {
      "bg": Object {
        "base": "#004",
        "disabled": "#003",
        "focused": "#005",
        "hovered": "#006",
        "selected": "#005",
      },
      "color": Object {
        "00": "#000",
        "10": "#001",
        "20": "#002",
        "30": "#003",
        "40": "#004",
        "50": "#005",
        "60": "#006",
        "70": "#007",
        "80": "#008",
        "90": "#009",
      },
      "fg": Object {
        "base": "#FF4",
        "disabled": "#FF3",
        "focused": "#FF5",
        "hovered": "#FF6",
        "selected": "#FF5",
      },
      "text": "#FF9",
    },
    "neutral": Object {
      "bg": Object {
        "base": "#FF4",
        "disabled": "#FF3",
        "focused": "#FF5",
        "hovered": "#FF6",
        "selected": "#FF5",
      },
      "color": Object {
        "00": "#FF0",
        "10": "#FF1",
        "20": "#FF2",
        "30": "#FF3",
        "40": "#FF4",
        "50": "#FF5",
        "60": "#FF6",
        "70": "#FF7",
        "80": "#FF8",
        "90": "#FF9",
      },
      "fg": Object {
        "base": "#FF4",
        "disabled": "#FF3",
        "focused": "#FF5",
        "hovered": "#FF6",
        "selected": "#FF5",
      },
      "text": "#008",
    },
    "positive": Object {
      "bg": Object {
        "base": "#FF4",
        "disabled": "#FF3",
        "focused": "#FF5",
        "hovered": "#FF6",
        "selected": "#FF5",
      },
      "color": Object {
        "00": "#FF0",
        "10": "#FF1",
        "20": "#FF2",
        "30": "#FF3",
        "40": "#FF4",
        "50": "#FF5",
        "60": "#FF6",
        "70": "#FF7",
        "80": "#FF8",
        "90": "#FF9",
      },
      "fg": Object {
        "base": "#FF4",
        "disabled": "#FF3",
        "focused": "#FF5",
        "hovered": "#FF6",
        "selected": "#FF5",
      },
      "text": "#FF8",
    },
    "primary": Object {
      "bg": Object {
        "base": "#BE4",
        "disabled": "#BE3",
        "focused": "#BE5",
        "hovered": "#BE6",
        "selected": "#BE5",
      },
      "color": Object {
        "00": "#BE0",
        "10": "#BE1",
        "20": "#BE2",
        "30": "#BE3",
        "40": "#BE4",
        "50": "#BE5",
        "60": "#BE6",
        "70": "#BE7",
        "80": "#BE8",
        "90": "#BE9",
      },
      "fg": Object {
        "base": "#FF4",
        "disabled": "#FF3",
        "focused": "#FF5",
        "hovered": "#FF6",
        "selected": "#FF5",
      },
      "text": "#BE8",
    },
    "secondary": Object {
      "bg": Object {
        "base": "#ED4",
        "disabled": "#ED3",
        "focused": "#ED5",
        "hovered": "#ED6",
        "selected": "#ED5",
      },
      "color": Object {
        "00": "#BE0",
        "10": "#BE1",
        "20": "#BE2",
        "30": "#BE3",
        "40": "#BE4",
        "50": "#BE5",
        "60": "#BE6",
        "70": "#BE7",
        "80": "#BE8",
        "90": "#BE9",
      },
      "fg": Object {
        "base": "#FF4",
        "disabled": "#FF3",
        "focused": "#FF5",
        "hovered": "#FF6",
        "selected": "#FF5",
      },
      "text": "#BE8",
    },
    "tertiary": Object {
      "bg": Object {
        "base": "#BE4",
        "disabled": "#BE3",
        "focused": "#BE5",
        "hovered": "#BE6",
        "selected": "#BE5",
      },
      "color": Object {
        "00": "#BE0",
        "10": "#BE1",
        "20": "#BE2",
        "30": "#BE3",
        "40": "#BE4",
        "50": "#BE5",
        "60": "#BE6",
        "70": "#BE7",
        "80": "#BE8",
        "90": "#BE9",
      },
      "fg": Object {
        "base": "#EE4",
        "disabled": "#EE3",
        "focused": "#EE5",
        "hovered": "#EE6",
        "selected": "#EE5",
      },
      "text": "#BE8",
    },
    "warning": Object {
      "bg": Object {
        "base": "#ED4",
        "disabled": "#ED3",
        "focused": "#ED5",
        "hovered": "#ED6",
        "selected": "#ED5",
      },
      "color": Object {
        "00": "#ED0",
        "10": "#ED1",
        "20": "#ED2",
        "30": "#ED3",
        "40": "#ED4",
        "50": "#ED5",
        "60": "#ED6",
        "70": "#ED7",
        "80": "#ED8",
        "90": "#ED9",
      },
      "fg": Object {
        "base": "#ED9",
        "disabled": "#ED0",
        "focused": "#ED8",
        "hovered": "#ED7",
        "selected": "#ED6",
      },
      "text": "#ED8",
    },
  },
}
`;

exports[`System themes supports theme extending 1`] = `
Object {
  "palette": Object {
    "brand": Object {
      "bg": Object {
        "base": "#BE4",
        "disabled": "#BE3",
        "focused": "#BE5",
        "hovered": "#BE6",
        "selected": "#BE5",
      },
      "color": Object {
        "00": "#BE0",
        "10": "#BE1",
        "20": "#BE2",
        "30": "#BE3",
        "40": "#BE4",
        "50": "#BE5",
        "60": "#BE6",
        "70": "#BE7",
        "80": "#BE8",
        "90": "#BE9",
      },
      "fg": Object {
        "base": "#BE4",
        "disabled": "#BE3",
        "focused": "#BE5",
        "hovered": "#BE6",
        "selected": "#BE5",
      },
      "text": "#BE8",
    },
    "danger": Object {
      "bg": Object {
        "base": "#ED9",
        "disabled": "#003",
        "focused": "#005",
        "hovered": "#006",
        "selected": "#005",
      },
      "color": Object {
        "00": "#ED0",
        "10": "#ED1",
        "20": "#ED2",
        "30": "#ED3",
        "40": "#ED4",
        "50": "#ED5",
        "60": "#ED6",
        "70": "#ED7",
        "80": "#ED8",
        "90": "#ED9",
      },
      "fg": Object {
        "base": "#ED0",
        "disabled": "#003",
        "focused": "#005",
        "hovered": "#006",
        "selected": "#005",
      },
      "text": "#ED8",
    },
    "failure": Object {
      "bg": Object {
        "base": "#BE4",
        "disabled": "#BE3",
        "focused": "#BE5",
        "hovered": "#BE6",
        "selected": "#BE5",
      },
      "color": Object {
        "00": "#BE0",
        "10": "#BE1",
        "20": "#BE2",
        "30": "#BE3",
        "40": "#BE4",
        "50": "#BE5",
        "60": "#BE6",
        "70": "#BE7",
        "80": "#BE8",
        "90": "#BE9",
      },
      "fg": Object {
        "base": "#ff4",
        "disabled": "#ff3",
        "focused": "#ff5",
        "hovered": "#ff6",
        "selected": "#ff5",
      },
      "text": "#BE0",
    },
    "info": Object {
      "bg": Object {
        "base": "#ff4",
        "disabled": "#ff3",
        "focused": "#ff5",
        "hovered": "#ff6",
        "selected": "#ff5",
      },
      "color": Object {
        "00": "#ff0",
        "10": "#ff1",
        "20": "#ff2",
        "30": "#ff3",
        "40": "#ff4",
        "50": "#ff5",
        "60": "#ff6",
        "70": "#ff7",
        "80": "#ff8",
        "90": "#ff9",
      },
      "fg": Object {
        "base": "#004",
        "disabled": "#003",
        "focused": "#005",
        "hovered": "#006",
        "selected": "#005",
      },
      "text": "#ff8",
    },
    "muted": Object {
      "bg": Object {
        "base": "#004",
        "disabled": "#003",
        "focused": "#005",
        "hovered": "#006",
        "selected": "#005",
      },
      "color": Object {
        "00": "#000",
        "10": "#001",
        "20": "#002",
        "30": "#003",
        "40": "#004",
        "50": "#005",
        "60": "#006",
        "70": "#007",
        "80": "#008",
        "90": "#009",
      },
      "fg": Object {
        "base": "#ff4",
        "disabled": "#ff3",
        "focused": "#ff5",
        "hovered": "#ff6",
        "selected": "#ff5",
      },
      "text": "#008",
    },
    "neutral": Object {
      "bg": Object {
        "base": "#EE4",
        "disabled": "#EE3",
        "focused": "#EE5",
        "hovered": "#EE6",
        "selected": "#EE5",
      },
      "color": Object {
        "00": "#EE0",
        "10": "#EE1",
        "20": "#EE2",
        "30": "#EE3",
        "40": "#EE4",
        "50": "#EE5",
        "60": "#EE6",
        "70": "#EE7",
        "80": "#EE8",
        "90": "#EE9",
      },
      "fg": Object {
        "base": "#ff0",
        "disabled": "#ff1",
        "focused": "#ff4",
        "hovered": "#ff3",
        "selected": "#ff8",
      },
      "text": "#EE8",
    },
    "positive": Object {
      "bg": Object {
        "base": "#EE4",
        "disabled": "#EE3",
        "focused": "#EE5",
        "hovered": "#EE6",
        "selected": "#EE5",
      },
      "color": Object {
        "00": "#EE0",
        "10": "#EE1",
        "20": "#EE2",
        "30": "#EE3",
        "40": "#EE4",
        "50": "#EE5",
        "60": "#EE6",
        "70": "#EE7",
        "80": "#EE8",
        "90": "#EE9",
      },
      "fg": Object {
        "base": "#ff4",
        "disabled": "#ff3",
        "focused": "#ff5",
        "hovered": "#ff6",
        "selected": "#ff5",
      },
      "text": "#EE9",
    },
    "primary": Object {
      "bg": Object {
        "base": "#BE0",
        "disabled": "#BE1",
        "focused": "#005",
        "hovered": "#BE2",
        "selected": "#005",
      },
      "color": Object {
        "00": "#BE0",
        "10": "#BE1",
        "20": "#BE2",
        "30": "#BE3",
        "40": "#BE4",
        "50": "#BE5",
        "60": "#BE6",
        "70": "#BE7",
        "80": "#BE8",
        "90": "#BE9",
      },
      "fg": Object {
        "base": "#ff9",
        "disabled": "#003",
        "focused": "#ff8",
        "hovered": "#006",
        "selected": "#ff7",
      },
      "text": "#BE8",
    },
    "secondary": Object {
      "bg": Object {
        "base": "#ED4",
        "disabled": "#ED3",
        "focused": "#ED5",
        "hovered": "#ED6",
        "selected": "#ED5",
      },
      "color": Object {
        "00": "#ED0",
        "10": "#ED1",
        "20": "#ED2",
        "30": "#ED3",
        "40": "#ED4",
        "50": "#ED5",
        "60": "#ED6",
        "70": "#ED7",
        "80": "#ED8",
        "90": "#ED9",
      },
      "fg": Object {
        "base": "#ff4",
        "disabled": "#ff3",
        "focused": "#ff5",
        "hovered": "#ff6",
        "selected": "#ff5",
      },
      "text": "#ED8",
    },
    "tertiary": Object {
      "bg": Object {
        "base": "#BE4",
        "disabled": "#BE3",
        "focused": "#BE5",
        "hovered": "#BE6",
        "selected": "#BE5",
      },
      "color": Object {
        "00": "#BE0",
        "10": "#BE1",
        "20": "#BE2",
        "30": "#BE3",
        "40": "#BE4",
        "50": "#BE5",
        "60": "#BE6",
        "70": "#BE7",
        "80": "#BE8",
        "90": "#BE9",
      },
      "fg": Object {
        "base": "#BE4",
        "disabled": "#BE3",
        "focused": "#BE5",
        "hovered": "#BE6",
        "selected": "#BE5",
      },
      "text": "#BE8",
    },
    "warning": Object {
      "bg": Object {
        "base": "#ED4",
        "disabled": "#ED3",
        "focused": "#ED5",
        "hovered": "#ED6",
        "selected": "#ED5",
      },
      "color": Object {
        "00": "#ED0",
        "10": "#ED1",
        "20": "#ED2",
        "30": "#ED3",
        "40": "#ED4",
        "50": "#ED5",
        "60": "#ED6",
        "70": "#ED7",
        "80": "#ED8",
        "90": "#ED9",
      },
      "fg": Object {
        "base": "#ff9",
        "disabled": "#ff9",
        "focused": "#ff9",
        "hovered": "#ff9",
        "selected": "#ff9",
      },
      "text": "#ED8",
    },
  },
}
`;
